justfile

```diff
@@ -0,0 +1,253 @@
+rs-check-dirs:
+    #!/usr/bin/env bash
+    for toml in $(find . -mindepth 2 -name Cargo.toml | sort -r); do
+        d=${toml%/*}
+        echo "cd $d && {{ _cargo }} check"
+        (cd $d && {{ _cargo }} check --frozen \
+            {{ if build-type == "release" { "--release" } else { "" } }} \
+            {{ _features }} \
+            {{ _fmt }})
+        echo "cd $d && {{ _cargo }} check --tests"
+        (cd $d && {{ _cargo }} check --frozen --tests \
+            {{ if build-type == "release" { "--release" } else { "" } }} \
+            {{ _features }} \
+            {{ _fmt }})
+    done
+
+##
+## Policy Integration Tests
+##
+
+_controller-image := DOCKER_REGISTRY + "/controller"
+_proxy-image := DOCKER_REGISTRY + "/proxy"
+_proxy-init-image := "ghcr.io/linkerd/proxy-init"
+_policy-controller-image := DOCKER_REGISTRY + "/policy-controller"
+
+# Run the policy controller integration tests in a k3d cluster
+policy-test: test-cluster-install-linkerd && _policy-test-uninstall
+    cd policy-test && {{ _cargo }} {{ _cargo-test }}
+
+# Delete all test namespaces and remove Linkerd from the cluster.
+policy-test-cleanup: && _policy-test-uninstall
+    {{ _kubectl }} delete ns -l linkerd-policy-test
+
+# Install Linkerd on the test cluster using test images.
+test-cluster-install-linkerd: test-cluster-install-crds _policy-test-images
+    {{ _linkerd }} install \
+            --set 'imagePullPolicy=Never' \
+            --set 'controllerImage={{ _controller-image }}' \
+            --set 'linkerdVersion={{ image-tag }}' \
+            --set 'proxy.image.name={{ _proxy-image }}' \
+            --set 'proxy.image.version={{ image-tag }}' \
+            --set 'proxyInit.image.name={{ _proxy-init-image }}' \
+            --set "proxyInit.image.version=$(yq .proxyInit.image.version <charts/linkerd-control-plane/values.yaml)" \
+            --set 'policyController.image.name={{ _policy-controller-image }}' \
+            --set 'policyController.logLevel=info\,linkerd=trace\,kubert=trace' \
+        | {{ _kubectl }} apply -f -
+    {{ _linkerd }} check -o short --wait=1m
+
+# Install CRDs on the test cluster.
+test-cluster-install-crds: _test-cluster-exists && _test-cluster-crds-ready
+    {{ _linkerd }} install --crds | {{ _kubectl }} apply -f -
+
+_test-cluster-crds-ready:
+    {{ _kubectl }} wait --for condition=established --timeout=60s crd \
+        authorizationpolicies.policy.linkerd.io \
+        meshtlsauthentications.policy.linkerd.io \
+        networkauthentications.policy.linkerd.io \
+        serverauthorizations.policy.linkerd.io \
+        servers.policy.linkerd.io
+
+# Build/fetch the Linkerd containers and load them onto the test cluster.
+_policy-test-images: docker-pull-policy-test-deps docker-build-policy-controller && policy-test-load-images
+    bin/docker-build-controller
+    bin/docker-build-proxy
+
+docker-pull-policy-test-deps:
+    docker pull -q docker.io/bitnami/kubectl:latest
```

Comment:
Probably not important in the grand scheme of things but to make sure client & server versions match would it make sense to use `test-cluster-k8s` here?, e.g

`docker.io/bitnami/kubectl:{{test-cluster-k8s}}`

Or would it be better not to be so tightly coupled here?

Reply:
this is pre-pulling images referenced in the tests. So, we'd have to somehow configure that version in the tests, too... https://github.com/linkerd/linkerd2/blob/5d0ffa71b395c12fb4a23693fd0b80c4da17aecc/policy-test/src/curl.rs#L161

i think it's fine to just use `latest` unless/until we hit compatibility issues.